### PR TITLE
RTL and bidi support

### PR DESCRIPTION
**Pull request by idolize**
I've tested this code with sample Arabic and bi-directional text and verified it looks as it's supposed to.

I confirm I have signed the [Contributor License Agreement](https://docs.google.com/a/playcanvas.com/forms/d/1Ih69zQfJG-QDLIEpHr6CsaAs6fPORNOVnMv5nuo0cjk/viewform).

### CODE REVIEW

File: src/framework/components/element/component.js [hunk 1]
```diff
@@ -87,6 +87,7 @@ Object.assign(pc, function () {
      * @property {Number} materialAsset The id of the material asset to use when rendering an image. Only works for {@link pc.ELEMENTTYPE_IMAGE} types.
      * @property {pc.Material} material The material to use when rendering an image. Only works for {@link pc.ELEMENTTYPE_IMAGE} types.
      * @property {pc.Vec4} rect Specifies which region of the texture to use in order to render an image. Values range from 0 to 1 and indicate u, v, width, height. Only works for {@link pc.ELEMENTTYPE_IMAGE} types.
+     * @property {Function} rtlReorder Optional function to reorder the text for RTL languages.
```
Comment (vkalpias) [inline src/framework/components/element/component.js]: Are there going to be multiple ways of rtl reordering? If not I guess we could make this a boolean?

Comment (idolize) [reply src/framework/components/element/component.js]: The issue with making it a boolean is then we have to include the algorithm for RTL/bidi in the engine code itself, right?

Comment (vkalpias) [reply src/framework/components/element/component.js]: Yes we would. I don't know if that's a problem though I'm just curious. Is it like a huge library?

File: src/resources/font.js [hunk 2]
```diff
@@ -72,6 +72,17 @@ Object.assign(pc, function () {
 
         open: function (url, data, asset) {
             var font;
+            // convert v2 to v3 font data
```
Comment (daredevildave) [inline src/resources/font.js]: We might actually need to add a v1->v2 conversion here as well. v1 format is missing the `data.info.maps` section and the `map` property for each character.

Comment (daredevildave) [reply src/resources/font.js]: See Arial.json in the engine examples as a v1 sample

File: src/core/string.js [hunk 3]
```diff
@@ -26,18 +26,28 @@ pc.string = function () {
...
+                return { code: (first - HIGH_SURROGATE_BEGIN) * 0x400 + second - LOW_SURROGATE_BEGIN + 0x10000, long: true };
             }
         }
-        return first;
+        return { code: first, long: false };
+    }
+
+    function getCodePoint(string, i) {
+        var codePointData = getCodePointData(string, i);
+        return codePointData === undefined ? NaN : codePointData.code;
```
[the remark this answers is not in the document]
Comment (idolize) [reply src/core/string.js]: I did that just to ensure a `Number` type is always returned from the function, but I can change it to `null`

File: src/core/string.js [hunk 4]
```diff
@@ -26,18 +26,28 @@ pc.string = function () {
...
                 // https://mathiasbynens.be/notes/javascript-encoding#surrogate-formulae
-                return (first - HIGH_SURROGATE_BEGIN) * 0x400 + second - LOW_SURROGATE_BEGIN + 0x10000;
+                return { code: (first - HIGH_SURROGATE_BEGIN) * 0x400 + second - LOW_SURROGATE_BEGIN + 0x10000, long: true };
             }
         }
-        return first;
+        return { code: first, long: false };
+    }
+
+    function getCodePoint(string, i) {
```
Comment (daredevildave) [inline src/core/string.js]: We could probably move this down into the main body now?

**Comment by daredevildave** [comment]
Also, need to check the tests that are failing. Use 'npm run test' to check locally

File: src/core/string.js (hunk 4 above)
Comment (idolize) [reply src/core/string.js]: It's actually used in `isCodeBetween` right now, but I'll change that...

**Pull request merged**